// loder/karma.config.js
var webpack = require('webpack')
process.env.CHROME_BIN = require('puppeteer').executablePath()

module.exports = function(config) {
  config.set({
    files: [
      // all files ending in "test"
      './node_modules/phantomjs-polyfill/bind-polyfill.js',
      'test/test.js'
      // each file acts as entry point for the webpack configuration
    ],

    // frameworks to use
    frameworks: ['mocha'],

    preprocessors: {
      // only specify one entry point
      // and require all tests in there
      'test/test.js': ['webpack']
    },

    reporters: ['spec', 'coverage'],

    coverageReporter: {
      dir: 'build/coverage/',
      reporters: [{ type: 'html' }, { type: 'text' }, { type: 'text-summary' }]
    },

    webpack: {
      // webpack configuration
      module: {
        rules: [
          {
            test: /\.js$/,
            exclude: /node_modules/,
            use: [
              {
                loader: 'babel-loader',
                options: {
                  presets: ['es2015', 'stage-0'],
                  plugins: ['transform-runtime', 'istanbul']
                }
              }
            ]
          }
        ]
      }
    },

    webpackMiddleware: {
      // webpack-dev-middleware configuration
      noInfo: true
    },

    plugins: [
      require('karma-webpack'),
      require('istanbul-instrumenter-loader'),
      require('karma-mocha'),
      require('karma-coverage'),
      require('karma-phantomjs-launcher'),
      require('karma-chrome-launcher'),
      require('karma-spec-reporter')
    ],

    browsers: ['ChromeHeadless']
  })
}
